src/DataTypes/DataTypeDateTime64.h

```diff
@@ -45,40 +45,75 @@ class DataTypeDateTime64 final : public DataTypeDecimalBase<DateTime64>, public
     bool equals(const IDataType & rhs) const override;
 };
 
-
-/** Basic wrapper for Tansform-types for DateTime64.
+/** Transform-type wrapper for DateTime64, applies given Transform to DateTime64 value or only to a whole part of it.
  *
- * Allows reusing existing Transform (that takes DateTime-values as UInt32) with DateTime64-values,
- * by discarding fractional part and producing SAME return type as original Transform.
+ * Depending on what overloads of Transform::execute() are available, when called with DateTime64 value,
+ * invokes Transform::execute() with:
+ * * whole part of DateTime64 value, discarding fractional part.
+ * * DateTime64 value and scale factor.
  *
- * Such Transfotm-types are commonly used in Date/DateTime manipulation functions,
- * and implement static execute fucntion with following signature:
- *      R execute(UInt32, T, const DateLUTImpl &)
+ * Suitable Transfotm-types are commonly used in Date/DateTime manipulation functions,
+ * and should implement static (or const) fucntion with following signatures:
+ *      R execute(UInt32 whole_value, ... , const TimeZoneImpl &)
+ * OR
+ *      R execute(DateTime64 value, Int64 scale_factor, ... , const TimeZoneImpl &)
  *
  * Wehere R and T could be arbitrary types.
 */
 template <typename Transform>
-class DateTime64BasicTransformWrapper : public Transform
+class TransformDateTime64 : public Transform
 {
+private:
+    // Name<X>::name is X::name or default.
+    template<typename X>
+    struct Name
+    {
+        template<typename U, typename CHECK> struct SFINAE {};
+        template<typename U> static constexpr auto getName(SFINAE<U, decltype(U::name)>*) { return U::name; }
```

Comment:
It looks old-fashioned, maybe it's possible to use std::enable_if_t?

Reply:
fixed